### PR TITLE
optimize partition balance to make partition distribution more even

### LINKED ISSUE #2047: [Enhancement]: 分区均衡过程中尽量随机打散分区
# 现有问题
1. 非分区表聚合之后取消聚合，分区均衡后，同一个database上的所有非分区表都聚集到同一日志流上
  * 创建3个数据库，每个数据库包含100个非分区表，创建后每个数据库的非分区表被均匀地打散到各日志流上
```
SELECT ls_id, database_name, role, count(*) FROM oceanbase.DBA_OB_TABLE_LOCATIONS where table_id >= 500000 and role = 'LEADER' group by ls_id, database_name order by database_name;
+-------+---------------+--------+----------+
| ls_id | database_name | role   | count(*) |
+-------+---------------+--------+----------+
|  1001 | db_0          | LEADER |       34 |
|  1002 | db_0          | LEADER |       33 |
|  1003 | db_0          | LEADER |       33 |
|  1001 | db_1          | LEADER |       33 |
|  1002 | db_1          | LEADER |       33 |
|  1003 | db_1          | LEADER |       34 |
|  1001 | db_2          | LEADER |       33 |
|  1002 | db_2          | LEADER |       34 |
|  1003 | db_2          | LEADER |       33 |
+-------+---------------+--------+----------+
9 rows in set (0.295 sec)
```
  * 将db_0中的所有非分区表绑定到表组中，然后取消绑定
```
SELECT ls_id, database_name, role, count(*) FROM oceanbase.DBA_OB_TABLE_LOCATIONS where table_id >= 500000 and role = 'LEADER' group by ls_id, database_name order by database_name;
+-------+---------------+--------+----------+
| ls_id | database_name | role   | count(*) |
+-------+---------------+--------+----------+
|  1001 | db_0          | LEADER |      100 |
|  1001 | db_1          | LEADER |       33 |
|  1002 | db_1          | LEADER |       33 |
|  1003 | db_1          | LEADER |       34 |
|  1001 | db_2          | LEADER |       33 |
|  1002 | db_2          | LEADER |       34 |
|  1003 | db_2          | LEADER |       33 |
+-------+---------------+--------+----------+
SELECT ls_id, database_name, role, count(*) FROM oceanbase.DBA_OB_TABLE_LOCATIONS where table_id >= 500000 and role = 'LEADER' group by ls_id, database_name order by database_name;
```
```
+-------+---------------+--------+----------+
| ls_id | database_name | role   | count(*) |
+-------+---------------+--------+----------+
|  1001 | db_0          | LEADER |      100 |
|  1002 | db_1          | LEADER |       53 |
|  1003 | db_1          | LEADER |       47 |
|  1002 | db_2          | LEADER |       47 |
|  1003 | db_2          | LEADER |       53 |
+-------+---------------+--------+----------+
```
虽然分区数目在各日志流之间均衡，但是db_0的所有非分区表聚合到日志流1001中
2. 分区表聚合后再重新打散，分区均衡后，分区在各日志流上变得连续
* primary_zone：random -> zone1 -> random
```
obclient [db_3]> SELECT ls_id, partition_name, role FROM oceanbase.DBA_OB_TABLE_LOCATIONS where table_name = 'table_part1'  and role = 'LEADER' order by ls_id;
+-------+----------------+--------+
| ls_id | partition_name | role   |
+-------+----------------+--------+
|  1001 | p1             | LEADER |
|  1001 | p2             | LEADER |
|  1001 | p3             | LEADER |
|  1001 | p4             | LEADER |
|  1001 | p5             | LEADER |
|  1001 | p6             | LEADER |
|  1001 | p7             | LEADER |
|  1001 | p8             | LEADER |
|  1001 | p9             | LEADER |
|  1001 | p10            | LEADER |
|  1001 | p11            | LEADER |
|  1001 | p12            | LEADER |
|  1001 | p13            | LEADER |
|  1001 | p14            | LEADER |
|  1001 | p15            | LEADER |
|  1001 | p16            | LEADER |
|  1001 | p17            | LEADER |
|  1001 | p18            | LEADER |
|  1001 | p19            | LEADER |
|  1001 | p20            | LEADER |
|  1001 | p21            | LEADER |
|  1001 | p22            | LEADER |
|  1001 | p23            | LEADER |
|  1001 | p24            | LEADER |
|  1001 | p25            | LEADER |
|  1001 | p26            | LEADER |
|  1001 | p27            | LEADER |
|  1001 | p28            | LEADER |
|  1001 | p29            | LEADER |
|  1001 | p30            | LEADER |
|  1001 | p31            | LEADER |
|  1001 | p32            | LEADER |
|  1001 | p33            | LEADER |
|  1001 | p0             | LEADER |
|  1004 | p98            | LEADER |
|  1004 | p66            | LEADER |
|  1004 | p67            | LEADER |
|  1004 | p68            | LEADER |
|  1004 | p69            | LEADER |
|  1004 | p70            | LEADER |
|  1004 | p71            | LEADER |
|  1004 | p72            | LEADER |
|  1004 | p73            | LEADER |
|  1004 | p74            | LEADER |
|  1004 | p75            | LEADER |
|  1004 | p76            | LEADER |
|  1004 | p77            | LEADER |
|  1004 | p78            | LEADER |
|  1004 | p79            | LEADER |
|  1004 | p80            | LEADER |
|  1004 | p81            | LEADER |
|  1004 | p82            | LEADER |
|  1004 | p83            | LEADER |
|  1004 | p84            | LEADER |
|  1004 | p85            | LEADER |
|  1004 | p86            | LEADER |
|  1004 | p87            | LEADER |
|  1004 | p88            | LEADER |
|  1004 | p89            | LEADER |
|  1004 | p90            | LEADER |
|  1004 | p91            | LEADER |
|  1004 | p92            | LEADER |
|  1004 | p93            | LEADER |
|  1004 | p94            | LEADER |
|  1004 | p95            | LEADER |
|  1004 | p96            | LEADER |
|  1004 | p97            | LEADER |
|  1005 | p50            | LEADER |
|  1005 | p37            | LEADER |
|  1005 | p38            | LEADER |
|  1005 | p39            | LEADER |
|  1005 | p40            | LEADER |
|  1005 | p41            | LEADER |
|  1005 | p42            | LEADER |
|  1005 | p43            | LEADER |
|  1005 | p44            | LEADER |
|  1005 | p45            | LEADER |
|  1005 | p46            | LEADER |
|  1005 | p47            | LEADER |
|  1005 | p48            | LEADER |
|  1005 | p49            | LEADER |
|  1005 | p34            | LEADER |
|  1005 | p51            | LEADER |
|  1005 | p52            | LEADER |
|  1005 | p53            | LEADER |
|  1005 | p54            | LEADER |
|  1005 | p55            | LEADER |
|  1005 | p56            | LEADER |
|  1005 | p57            | LEADER |
|  1005 | p58            | LEADER |
|  1005 | p59            | LEADER |
|  1005 | p60            | LEADER |
|  1005 | p61            | LEADER |
|  1005 | p62            | LEADER |
|  1005 | p63            | LEADER |
|  1005 | p64            | LEADER |
|  1005 | p65            | LEADER |
|  1005 | p35            | LEADER |
|  1005 | p36            | LEADER |
|  1005 | p99            | LEADER |
+-------+----------------+--------+
100 rows in set (1.528 sec)
```
# 期望改进
1. 所有非分区表分区聚合再打散后，同一数据库下的所有非分区表打散到各日志流上
2. 分区表的分区聚合再打散后，分区在各日志流上尽可能随机分布

### PR DESCRIPTION
**Pull request by Young-Cody**
### Task Description
close #2047

### Passed Regressions

1. 所有非分区表分区聚合再打散后，同一数据库下的所有非分区表打散到各日志流上
```
SELECT ls_id, database_name, role, count(*) FROM oceanbase.DBA_OB_TABLE_LOCATIONS where table_id >= 500000 and partition_name = 'NULL' and role = 'LEADER' group by ls_i
d, database_name;
+-------+---------------+--------+----------+
| ls_id | database_name | role   | count(*) |
+-------+---------------+--------+----------+
|  1001 | db_0          | LEADER |       33 |
|  1004 | db_0          | LEADER |       34 |
|  1006 | db_0          | LEADER |       33 |
|  1006 | db_1          | LEADER |       34 |
|  1004 | db_1          | LEADER |       33 |
|  1001 | db_1          | LEADER |       33 |
|  1001 | db_2          | LEADER |       34 |
|  1004 | db_2          | LEADER |       33 |
|  1006 | db_2          | LEADER |       33 |
+-------+---------------+--------+----------+
9 rows in set (0.361 sec)
```
2. 分区表的分区聚合再打散后，分区在各日志流上尽可能随机分布
```
SELECT ls_id, table_name, object_id, partition_name, subpartition_name FROM oceanbase.DBA_OB_TABLE_LOCATIONS where table_type = 'user table'  and role = 'LEADER' and table_name = 'table_part1'  order by ls_id, object_id;
+-------+-------------+-----------+----------------+-------------------+
| ls_id | table_name  | object_id | partition_name | subpartition_name |
+-------+-------------+-----------+----------------+-------------------+
|  1001 | table_part1 |    500317 | p1             | NULL              |
|  1001 | table_part1 |    500320 | p4             | NULL              |
|  1001 | table_part1 |    500323 | p7             | NULL              |
|  1001 | table_part1 |    500326 | p10            | NULL              |
|  1001 | table_part1 |    500329 | p13            | NULL              |
|  1001 | table_part1 |    500332 | p16            | NULL              |
|  1001 | table_part1 |    500335 | p19            | NULL              |
|  1001 | table_part1 |    500338 | p22            | NULL              |
|  1001 | table_part1 |    500341 | p25            | NULL              |
|  1001 | table_part1 |    500344 | p28            | NULL              |
|  1001 | table_part1 |    500347 | p31            | NULL              |
|  1001 | table_part1 |    500350 | p34            | NULL              |
|  1001 | table_part1 |    500353 | p37            | NULL              |
|  1001 | table_part1 |    500356 | p40            | NULL              |
|  1001 | table_part1 |    500359 | p43            | NULL              |
|  1001 | table_part1 |    500362 | p46            | NULL              |
|  1001 | table_part1 |    500365 | p49            | NULL              |
|  1001 | table_part1 |    500368 | p52            | NULL              |
|  1001 | table_part1 |    500371 | p55            | NULL              |
|  1001 | table_part1 |    500374 | p58            | NULL              |
|  1001 | table_part1 |    500377 | p61            | NULL              |
|  1001 | table_part1 |    500380 | p64            | NULL              |
|  1001 | table_part1 |    500383 | p67            | NULL              |
|  1001 | table_part1 |    500386 | p70            | NULL              |
|  1001 | table_part1 |    500389 | p73            | NULL              |
|  1001 | table_part1 |    500392 | p76            | NULL              |
|  1001 | table_part1 |    500395 | p79            | NULL              |
|  1001 | table_part1 |    500398 | p82            | NULL              |
|  1001 | table_part1 |    500401 | p85            | NULL              |
|  1001 | table_part1 |    500404 | p88            | NULL              |
|  1001 | table_part1 |    500407 | p91            | NULL              |
|  1001 | table_part1 |    500410 | p94            | NULL              |
|  1001 | table_part1 |    500413 | p97            | NULL              |
|  1001 | table_part1 |    500416 | p100           | NULL              |
|  1001 | table_part1 |    500419 | p103           | NULL              |
|  1001 | table_part1 |    500422 | p106           | NULL              |
|  1001 | table_part1 |    500425 | p109           | NULL              |
|  1001 | table_part1 |    500428 | p112           | NULL              |
|  1001 | table_part1 |    500431 | p115           | NULL              |
|  1001 | table_part1 |    500434 | p118           | NULL              |
|  1001 | table_part1 |    500437 | p121           | NULL              |
|  1001 | table_part1 |    500440 | p124           | NULL              |
|  1001 | table_part1 |    500443 | p127           | NULL              |
|  1001 | table_part1 |    500446 | p130           | NULL              |
|  1001 | table_part1 |    500449 | p133           | NULL              |
|  1001 | table_part1 |    500452 | p136           | NULL              |
|  1001 | table_part1 |    500455 | p139           | NULL              |
|  1001 | table_part1 |    500458 | p142           | NULL              |
|  1001 | table_part1 |    500461 | p145           | NULL              |
|  1001 | table_part1 |    500464 | p148           | NULL              |
|  1001 | table_part1 |    500467 | p151           | NULL              |
|  1001 | table_part1 |    500470 | p154           | NULL              |
|  1001 | table_part1 |    500473 | p157           | NULL              |
|  1001 | table_part1 |    500476 | p160           | NULL              |
|  1001 | table_part1 |    500479 | p163           | NULL              |
|  1001 | table_part1 |    500482 | p166           | NULL              |
|  1001 | table_part1 |    500485 | p169           | NULL              |
|  1001 | table_part1 |    500488 | p172           | NULL              |
|  1001 | table_part1 |    500491 | p175           | NULL              |
|  1001 | table_part1 |    500494 | p178           | NULL              |
|  1001 | table_part1 |    500497 | p181           | NULL              |
|  1001 | table_part1 |    500500 | p184           | NULL              |
|  1001 | table_part1 |    500503 | p187           | NULL              |
|  1001 | table_part1 |    500506 | p190           | NULL              |
|  1001 | table_part1 |    500509 | p193           | NULL              |
|  1001 | table_part1 |    500512 | p196           | NULL              |
|  1001 | table_part1 |    500515 | p199           | NULL              |
|  1001 | table_part1 |    500518 | p202           | NULL              |
|  1001 | table_part1 |    500521 | p205           | NULL              |
|  1001 | table_part1 |    500524 | p208           | NULL              |
|  1001 | table_part1 |    500527 | p211           | NULL              |
|  1001 | table_part1 |    500530 | p214           | NULL              |
|  1001 | table_part1 |    500533 | p217           | NULL              |
|  1001 | table_part1 |    500536 | p220           | NULL              |
|  1001 | table_part1 |    500539 | p223           | NULL              |
|  1001 | table_part1 |    500542 | p226           | NULL              |
|  1001 | table_part1 |    500545 | p229           | NULL              |
|  1001 | table_part1 |    500548 | p232           | NULL              |
|  1001 | table_part1 |    500551 | p235           | NULL              |
|  1001 | table_part1 |    500554 | p238           | NULL              |
|  1001 | table_part1 |    500557 | p241           | NULL              |
|  1001 | table_part1 |    500560 | p244           | NULL              |
|  1001 | table_part1 |    500563 | p247           | NULL              |
|  1001 | table_part1 |    500566 | p250           | NULL              |
|  1001 | table_part1 |    500569 | p253           | NULL              |
|  1001 | table_part1 |    500572 | p256           | NULL              |
|  1001 | table_part1 |    500575 | p259           | NULL              |
|  1001 | table_part1 |    500578 | p262           | NULL              |
|  1001 | table_part1 |    500581 | p265           | NULL              |
|  1001 | table_part1 |    500584 | p268           | NULL              |
|  1001 | table_part1 |    500587 | p271           | NULL              |
|  1001 | table_part1 |    500590 | p274           | NULL              |
|  1001 | table_part1 |    500593 | p277           | NULL              |
|  1001 | table_part1 |    500596 | p280           | NULL              |
|  1001 | table_part1 |    500599 | p283           | NULL              |
|  1001 | table_part1 |    500602 | p286           | NULL              |
|  1001 | table_part1 |    500605 | p289           | NULL              |
|  1001 | table_part1 |    500608 | p292           | NULL              |
|  1001 | table_part1 |    500611 | p295           | NULL              |
|  1001 | table_part1 |    500614 | p298           | NULL              |
|  1004 | table_part1 |    500318 | p2             | NULL              |
|  1004 | table_part1 |    500321 | p5             | NULL              |
|  1004 | table_part1 |    500324 | p8             | NULL              |
|  1004 | table_part1 |    500327 | p11            | NULL              |
|  1004 | table_part1 |    500330 | p14            | NULL              |
|  1004 | table_part1 |    500333 | p17            | NULL              |
|  1004 | table_part1 |    500336 | p20            | NULL              |
|  1004 | table_part1 |    500339 | p23            | NULL              |
|  1004 | table_part1 |    500342 | p26            | NULL              |
|  1004 | table_part1 |    500345 | p29            | NULL              |
|  1004 | table_part1 |    500348 | p32            | NULL              |
|  1004 | table_part1 |    500351 | p35            | NULL              |
|  1004 | table_part1 |    500354 | p38            | NULL              |
|  1004 | table_part1 |    500357 | p41            | NULL              |
|  1004 | table_part1 |    500360 | p44            | NULL              |
|  1004 | table_part1 |    500363 | p47            | NULL              |
|  1004 | table_part1 |    500366 | p50            | NULL              |
|  1004 | table_part1 |    500369 | p53            | NULL              |
|  1004 | table_part1 |    500372 | p56            | NULL              |
|  1004 | table_part1 |    500375 | p59            | NULL              |
|  1004 | table_part1 |    500378 | p62            | NULL              |
|  1004 | table_part1 |    500381 | p65            | NULL              |
|  1004 | table_part1 |    500384 | p68            | NULL              |
|  1004 | table_part1 |    500387 | p71            | NULL              |
|  1004 | table_part1 |    500390 | p74            | NULL              |
|  1004 | table_part1 |    500393 | p77            | NULL              |
|  1004 | table_part1 |    500396 | p80            | NULL              |
|  1004 | table_part1 |    500399 | p83            | NULL              |
|  1004 | table_part1 |    500402 | p86            | NULL              |
|  1004 | table_part1 |    500405 | p89            | NULL              |
|  1004 | table_part1 |    500408 | p92            | NULL              |
|  1004 | table_part1 |    500411 | p95            | NULL              |
|  1004 | table_part1 |    500414 | p98            | NULL              |
|  1004 | table_part1 |    500417 | p101           | NULL              |
|  1004 | table_part1 |    500420 | p104           | NULL              |
|  1004 | table_part1 |    500423 | p107           | NULL              |
|  1004 | table_part1 |    500426 | p110           | NULL              |
|  1004 | table_part1 |    500429 | p113           | NULL              |
|  1004 | table_part1 |    500432 | p116           | NULL              |
|  1004 | table_part1 |    500435 | p119           | NULL              |
|  1004 | table_part1 |    500438 | p122           | NULL              |
|  1004 | table_part1 |    500441 | p125           | NULL              |
|  1004 | table_part1 |    500444 | p128           | NULL              |
|  1004 | table_part1 |    500447 | p131           | NULL              |
|  1004 | table_part1 |    500450 | p134           | NULL              |
|  1004 | table_part1 |    500453 | p137           | NULL              |
|  1004 | table_part1 |    500456 | p140           | NULL              |
|  1004 | table_part1 |    500459 | p143           | NULL              |
|  1004 | table_part1 |    500462 | p146           | NULL              |
|  1004 | table_part1 |    500465 | p149           | NULL              |
|  1004 | table_part1 |    500468 | p152           | NULL              |
|  1004 | table_part1 |    500471 | p155           | NULL              |
|  1004 | table_part1 |    500474 | p158           | NULL              |
|  1004 | table_part1 |    500477 | p161           | NULL              |
|  1004 | table_part1 |    500480 | p164           | NULL              |
|  1004 | table_part1 |    500483 | p167           | NULL              |
|  1004 | table_part1 |    500486 | p170           | NULL              |
|  1004 | table_part1 |    500489 | p173           | NULL              |
|  1004 | table_part1 |    500492 | p176           | NULL              |
|  1004 | table_part1 |    500495 | p179           | NULL              |
|  1004 | table_part1 |    500498 | p182           | NULL              |
|  1004 | table_part1 |    500501 | p185           | NULL              |
|  1004 | table_part1 |    500504 | p188           | NULL              |
|  1004 | table_part1 |    500507 | p191           | NULL              |
|  1004 | table_part1 |    500510 | p194           | NULL              |
|  1004 | table_part1 |    500513 | p197           | NULL              |
|  1004 | table_part1 |    500516 | p200           | NULL              |
|  1004 | table_part1 |    500519 | p203           | NULL              |
|  1004 | table_part1 |    500522 | p206           | NULL              |
|  1004 | table_part1 |    500525 | p209           | NULL              |
|  1004 | table_part1 |    500528 | p212           | NULL              |
|  1004 | table_part1 |    500531 | p215           | NULL              |
|  1004 | table_part1 |    500534 | p218           | NULL              |
|  1004 | table_part1 |    500537 | p221           | NULL              |
|  1004 | table_part1 |    500540 | p224           | NULL              |
|  1004 | table_part1 |    500543 | p227           | NULL              |
|  1004 | table_part1 |    500546 | p230           | NULL              |
|  1004 | table_part1 |    500549 | p233           | NULL              |
|  1004 | table_part1 |    500552 | p236           | NULL              |
|  1004 | table_part1 |    500555 | p239           | NULL              |
|  1004 | table_part1 |    500558 | p242           | NULL              |
|  1004 | table_part1 |    500561 | p245           | NULL              |
|  1004 | table_part1 |    500564 | p248           | NULL              |
|  1004 | table_part1 |    500567 | p251           | NULL              |
|  1004 | table_part1 |    500570 | p254           | NULL              |
|  1004 | table_part1 |    500573 | p257           | NULL              |
|  1004 | table_part1 |    500576 | p260           | NULL              |
|  1004 | table_part1 |    500579 | p263           | NULL              |
|  1004 | table_part1 |    500582 | p266           | NULL              |
|  1004 | table_part1 |    500585 | p269           | NULL              |
|  1004 | table_part1 |    500588 | p272           | NULL              |
|  1004 | table_part1 |    500591 | p275           | NULL              |
|  1004 | table_part1 |    500594 | p278           | NULL              |
|  1004 | table_part1 |    500597 | p281           | NULL              |
|  1004 | table_part1 |    500600 | p284           | NULL              |
|  1004 | table_part1 |    500603 | p287           | NULL              |
|  1004 | table_part1 |    500606 | p290           | NULL              |
|  1004 | table_part1 |    500609 | p293           | NULL              |
|  1004 | table_part1 |    500612 | p296           | NULL              |
|  1004 | table_part1 |    500615 | p299           | NULL              |
|  1006 | table_part1 |    500316 | p0             | NULL              |
|  1006 | table_part1 |    500319 | p3             | NULL              |
|  1006 | table_part1 |    500322 | p6             | NULL              |
|  1006 | table_part1 |    500325 | p9             | NULL              |
|  1006 | table_part1 |    500328 | p12            | NULL              |
|  1006 | table_part1 |    500331 | p15            | NULL              |
|  1006 | table_part1 |    500334 | p18            | NULL              |
|  1006 | table_part1 |    500337 | p21            | NULL              |
|  1006 | table_part1 |    500340 | p24            | NULL              |
|  1006 | table_part1 |    500343 | p27            | NULL              |
|  1006 | table_part1 |    500346 | p30            | NULL              |
|  1006 | table_part1 |    500349 | p33            | NULL              |
|  1006 | table_part1 |    500352 | p36            | NULL              |
|  1006 | table_part1 |    500355 | p39            | NULL              |
|  1006 | table_part1 |    500358 | p42            | NULL              |
|  1006 | table_part1 |    500361 | p45            | NULL              |
|  1006 | table_part1 |    500364 | p48            | NULL              |
|  1006 | table_part1 |    500367 | p51            | NULL              |
|  1006 | table_part1 |    500370 | p54            | NULL              |
|  1006 | table_part1 |    500373 | p57            | NULL              |
|  1006 | table_part1 |    500376 | p60            | NULL              |
|  1006 | table_part1 |    500379 | p63            | NULL              |
|  1006 | table_part1 |    500382 | p66            | NULL              |
|  1006 | table_part1 |    500385 | p69            | NULL              |
|  1006 | table_part1 |    500388 | p72            | NULL              |
|  1006 | table_part1 |    500391 | p75            | NULL              |
|  1006 | table_part1 |    500394 | p78            | NULL              |
|  1006 | table_part1 |    500397 | p81            | NULL              |
|  1006 | table_part1 |    500400 | p84            | NULL              |
|  1006 | table_part1 |    500403 | p87            | NULL              |
|  1006 | table_part1 |    500406 | p90            | NULL              |
|  1006 | table_part1 |    500409 | p93            | NULL              |
|  1006 | table_part1 |    500412 | p96            | NULL              |
|  1006 | table_part1 |    500415 | p99            | NULL              |
|  1006 | table_part1 |    500418 | p102           | NULL              |
|  1006 | table_part1 |    500421 | p105           | NULL              |
|  1006 | table_part1 |    500424 | p108           | NULL              |
|  1006 | table_part1 |    500427 | p111           | NULL              |
|  1006 | table_part1 |    500430 | p114           | NULL              |
|  1006 | table_part1 |    500433 | p117           | NULL              |
|  1006 | table_part1 |    500436 | p120           | NULL              |
|  1006 | table_part1 |    500439 | p123           | NULL              |
|  1006 | table_part1 |    500442 | p126           | NULL              |
|  1006 | table_part1 |    500445 | p129           | NULL              |
|  1006 | table_part1 |    500448 | p132           | NULL              |
|  1006 | table_part1 |    500451 | p135           | NULL              |
|  1006 | table_part1 |    500454 | p138           | NULL              |
|  1006 | table_part1 |    500457 | p141           | NULL              |
|  1006 | table_part1 |    500460 | p144           | NULL              |
|  1006 | table_part1 |    500463 | p147           | NULL              |
|  1006 | table_part1 |    500466 | p150           | NULL              |
|  1006 | table_part1 |    500469 | p153           | NULL              |
|  1006 | table_part1 |    500472 | p156           | NULL              |
|  1006 | table_part1 |    500475 | p159           | NULL              |
|  1006 | table_part1 |    500478 | p162           | NULL              |
|  1006 | table_part1 |    500481 | p165           | NULL              |
|  1006 | table_part1 |    500484 | p168           | NULL              |
|  1006 | table_part1 |    500487 | p171           | NULL              |
|  1006 | table_part1 |    500490 | p174           | NULL              |
|  1006 | table_part1 |    500493 | p177           | NULL              |
|  1006 | table_part1 |    500496 | p180           | NULL              |
|  1006 | table_part1 |    500499 | p183           | NULL              |
|  1006 | table_part1 |    500502 | p186           | NULL              |
|  1006 | table_part1 |    500505 | p189           | NULL              |
|  1006 | table_part1 |    500508 | p192           | NULL              |
|  1006 | table_part1 |    500511 | p195           | NULL              |
|  1006 | table_part1 |    500514 | p198           | NULL              |
|  1006 | table_part1 |    500517 | p201           | NULL              |
|  1006 | table_part1 |    500520 | p204           | NULL              |
|  1006 | table_part1 |    500523 | p207           | NULL              |
|  1006 | table_part1 |    500526 | p210           | NULL              |
|  1006 | table_part1 |    500529 | p213           | NULL              |
|  1006 | table_part1 |    500532 | p216           | NULL              |
|  1006 | table_part1 |    500535 | p219           | NULL              |
|  1006 | table_part1 |    500538 | p222           | NULL              |
|  1006 | table_part1 |    500541 | p225           | NULL              |
|  1006 | table_part1 |    500544 | p228           | NULL              |
|  1006 | table_part1 |    500547 | p231           | NULL              |
|  1006 | table_part1 |    500550 | p234           | NULL              |
|  1006 | table_part1 |    500553 | p237           | NULL              |
|  1006 | table_part1 |    500556 | p240           | NULL              |
|  1006 | table_part1 |    500559 | p243           | NULL              |
|  1006 | table_part1 |    500562 | p246           | NULL              |
|  1006 | table_part1 |    500565 | p249           | NULL              |
|  1006 | table_part1 |    500568 | p252           | NULL              |
|  1006 | table_part1 |    500571 | p255           | NULL              |
|  1006 | table_part1 |    500574 | p258           | NULL              |
|  1006 | table_part1 |    500577 | p261           | NULL              |
|  1006 | table_part1 |    500580 | p264           | NULL              |
|  1006 | table_part1 |    500583 | p267           | NULL              |
|  1006 | table_part1 |    500586 | p270           | NULL              |
|  1006 | table_part1 |    500589 | p273           | NULL              |
|  1006 | table_part1 |    500592 | p276           | NULL              |
|  1006 | table_part1 |    500595 | p279           | NULL              |
|  1006 | table_part1 |    500598 | p282           | NULL              |
|  1006 | table_part1 |    500601 | p285           | NULL              |
|  1006 | table_part1 |    500604 | p288           | NULL              |
|  1006 | table_part1 |    500607 | p291           | NULL              |
|  1006 | table_part1 |    500610 | p294           | NULL              |
|  1006 | table_part1 |    500613 | p297           | NULL              |
+-------+-------------+-----------+----------------+-------------------+
300 rows in set (0.531 sec)
```

### CODE REVIEW

File: src/rootserver/balance/ob_balance_group_info.h [hunk 1]
```diff
@@ -55,60 +56,139 @@ class ObTransferPartGroup
...
+    alloc_(alloc),
+    part_group_buckets_(OB_MALLOC_NORMAL_BLOCK_SIZE, ModulePageAllocator(alloc, "PGBuckets")),
+    part_group_cnt_(0)
+  {}
+  ~ObBalanceGroupUnit();
+  int init(const int64_t bucket_num);
+  bool is_valid() const { return inited_; }
+  int64_t get_part_group_count() const { return part_group_cnt_; }
+  int append_part_group(const uint64_t part_group_uid, ObTransferPartGroup *const part_group);
+  int append_part_group(const int64_t bucket_idx, ObTransferPartGroup *const part_group);
```
Comment (ZhenNan0) [inline src/rootserver/balance/ob_balance_group_info.h:75]: 1.Functions that are not exposed to the public need to be set to private.
2.Try not to set the same function name, it is easy to be misused.

File: src/rootserver/ob_partition_balance.cpp [hunk 2]
```diff
@@ -242,85 +242,66 @@ int ObPartitionBalance::on_new_partition(
...
-  ObArray<ObLSPartGroupDesc *> *ls_part_desc_arr = NULL;
-  ls_part_desc_arr = bg_map_.get(bg);
-  if (OB_ISNULL(ls_part_desc_arr)) {
-    ObArray<ObLSPartGroupDesc *> ls_part_desc_arr_obj;
-    if (OB_FAIL(bg_map_.set_refactored(bg, ls_part_desc_arr_obj))) {
+  ObArray<ObBalanceGroupInfo *> *bg_ls_array = NULL;
+  bg_ls_array = bg_map_.get(const_cast<ObBalanceGroup&>(bg));
+  if (OB_ISNULL(bg_ls_array)) {
+    ObArray<ObBalanceGroupInfo *> bg_ls_array_obj;
+    const int64_t bucket_num =  bg.is_non_part_table_bg() ? 1 : ls_desc_array_.count();
```
Comment (ZhenNan0) [inline src/rootserver/ob_partition_balance.cpp:263]: bucket_num is bound to bg_id, no need to specify this parameter externally

File: src/rootserver/ob_partition_balance.cpp [hunk 3]
```diff
@@ -242,85 +242,66 @@ int ObPartitionBalance::on_new_partition(
...
+    const ObTransferPartInfo &part_info, const int64_t tablet_size)
 {
   int ret = OB_SUCCESS;
-  ObArray<ObLSPartGroupDesc *> *ls_part_desc_arr = NULL;
-  ls_part_desc_arr = bg_map_.get(bg);
-  if (OB_ISNULL(ls_part_desc_arr)) {
-    ObArray<ObLSPartGroupDesc *> ls_part_desc_arr_obj;
-    if (OB_FAIL(bg_map_.set_refactored(bg, ls_part_desc_arr_obj))) {
+  ObArray<ObBalanceGroupInfo *> *bg_ls_array = NULL;
+  bg_ls_array = bg_map_.get(const_cast<ObBalanceGroup&>(bg));
```
[the remark this answers is not in the document]
Comment (Young-Cody) [reply src/rootserver/ob_partition_balance.cpp:260]: Is the use of const_cast prohibited by coding specifications?

File: src/rootserver/balance/ob_balance_group_info.cpp [hunk 4]
```diff
@@ -34,62 +34,323 @@ int ObTransferPartGroup::add_part(const ObTransferPartInfo &part, int64_t data_s
...
+  double ratio_max = -DBL_MAX;
+  bucket_idx = OB_INVALID_INDEX;
+  for (int64_t i = 0; i < dst_unit.part_group_buckets_.count(); i++) {
+    const ObArray<ObTransferPartGroup*> &src_part_groups = part_group_buckets_.at(i);
+    const ObArray<ObTransferPartGroup*> &dst_part_groups = dst_unit.part_group_buckets_.at(i);
+    int64_t src_cnt = src_part_groups.count();
+    int64_t dst_cnt = dst_part_groups.count();
+    if (src_cnt != 0) {
+      double ratio = static_cast<double>(dst_cnt) / static_cast<double>(src_cnt);
+      if (OB_INVALID_INDEX == bucket_idx || ratio > ratio_max) {
```
[the remark this answers is not in the document]
Comment (Young-Cody) [reply src/rootserver/balance/ob_balance_group_info.cpp:255]: I still don't understand why it's incorrect

File: src/rootserver/ob_partition_balance.cpp [hunk 5]
```diff
@@ -239,88 +250,84 @@ int ObPartitionBalance::on_new_partition(
...
-        if (OB_FAIL(ls_part_desc_arr->at(idx)->add_new_part_group(part_group))) {
-          LOG_WARN("add_new_part_group failed", KR(ret), K(ls_id), K(idx), KPC(ls_part_desc_arr), K(part_group));
+    ARRAY_FOREACH(*bg_ls_array, idx) {
+      ObBalanceGroupInfo *bg_info = bg_ls_array->at(idx);
+      if (OB_ISNULL(bg_info)) {
+        ret = OB_ERR_UNEXPECTED;
+        LOG_WARN("bg_info is null", KR(ret), K(bg_info), KPC(bg_ls_array), K(idx));
+      } else if (bg_info->get_ls_id() == ls_id) {
+        if (OB_FAIL(bg_info->append_part(bg_unit_id, part_group_uid,
+                                                      part_info, tablet_size))) {
```
Comment (ZhenNan0) [inline src/rootserver/ob_partition_balance.cpp:320]: out of alignment

File: src/rootserver/balance/ob_balance_group_define.h [hunk 6]
```diff
@@ -58,6 +58,13 @@ struct ObBalanceGroupID
   bool operator !=(const ObBalanceGroupID &other) const {
     return !(*this == other);
   }
+  bool is_non_part_table_bg() const {
+    return 0 == id_high_ && 0 == id_low_;
+  }
+  void reset() {
+    id_high_ = common::OB_INVALID_ID;
+    id_low_ = common::OB_INVALID_ID;
```
Comment (Young-Cody) [inline src/rootserver/balance/ob_balance_group_define.h:66]: 定义常量

File: src/rootserver/ob_ls_balance_helper.cpp [hunk 7]
```diff
@@ -850,34 +869,41 @@ int ObLSBalanceTaskHelper::construct_ls_merge_task_(
   return ret;
 }
 
-int ObLSBalanceTaskHelper::construct_ls_part_info_(const ObSplitLSParam &src_ls, ObTransferPartList &part_list)
+int ObLSBalanceTaskHelper::construct_ls_part_info_(
+    const ObSplitLSParam &src_ls,
+    const share::ObLSID &dest_ls_id,
```
Comment (Young-Cody) [inline src/rootserver/ob_ls_balance_helper.cpp:874]: 命名有误导，修改，注释举例子

File: src/rootserver/balance/ob_tenant_ls_balance_group_info.h [hunk 8]
```diff
@@ -80,6 +82,7 @@ class ObTenantLSBalanceGroupInfo final : public ObAllBalanceGroupBuilder::NewPar
   // map for all balance groups on tenant every LS
   // If LS is empty, it does not exist in this map
   common::hash::ObHashMap<share::ObLSID, ObLSBalanceGroupInfo *> ls_bg_map_;
+  int64_t ls_num_;
```
Comment (Young-Cody) [inline src/rootserver/balance/ob_tenant_ls_balance_group_info.h:85]: 重命名

Comment (Young-Cody) [inline src/rootserver/balance/ob_tenant_ls_balance_group_info.h:85]: 重命名

**Pull request closed without merging**